File: flux/apps/games/futboss-ai/apps/cli/pkg/models/player.go
// FutBoss AI - Player Model (Go)
// Author: Bruno Lucena (bruno@lucena.cloud)

package models

type Position string

const (
	PositionGK  Position = "GK"
	PositionCB  Position = "CB"
	PositionLB  Position = "LB"
	PositionRB  Position = "RB"
	PositionCDM Position = "CDM"
	PositionCM  Position = "CM"
	PositionCAM Position = "CAM"
	PositionLW  Position = "LW"
	PositionRW  Position = "RW"
	PositionST  Position = "ST"
)

type Temperament string

const (
	TemperamentCalm       Temperament = "calm"
	TemperamentExplosive  Temperament = "explosive"
	TemperamentCalculated Temperament = "calculated"
)

type PlayStyle string

const (
	PlayStyleOffensive PlayStyle = "offensive"
	PlayStyleDefensive PlayStyle = "defensive"
	PlayStyleBalanced  PlayStyle = "balanced"
)

type PlayerAttributes struct {
	Speed       int `json:"speed"`
	Strength    int `json:"strength"`
	Stamina     int `json:"stamina"`
	Finishing   int `json:"finishing"`
	Passing     int `json:"passing"`
	Dribbling   int `json:"dribbling"`
	Defense     int `json:"defense"`
	Intelligence int `json:"intelligence"`
	Aggression  int `json:"aggression"`
	Leadership  int `json:"leadership"`
	Creativity  int `json:"creativity"`
}

func (a *PlayerAttributes) Overall() int {
	attrs := []int{
		a.Speed, a.Strength, a.Stamina,
		a.Finishing, a.Passing, a.Dribbling, a.Defense,
		a.Intelligence, a.Aggression, a.Leadership, a.Creativity,
	}
	sum := 0
	for _, v := range attrs {
		sum += v
	}
	return sum / len(attrs)
}

type PlayerPersonality struct {
	Temperament Temperament `json:"temperament"`
	PlayStyle   PlayStyle   `json:"play_style"`
}

type Player struct {
	ID          string            `json:"id"`
	Name        string            `json:"name"`
	Position    Position          `json:"position"`
	Nationality string            `json:"nationality"`
	Age         int               `json:"age"`
	Attributes  PlayerAttributes  `json:"attributes"`
	Personality PlayerPersonality `json:"personality"`
	TeamID      *string           `json:"team_id,omitempty"`
	Price       int               `json:"price"`
	IsListed    bool              `json:"is_listed"`
}

func (p *Player) GetMarketValue() int {
	baseValue := p.Attributes.Overall() * 10
	ageModifier := 1.0

	if p.Age < 23 {
		ageModifier = 1.3
	} else if p.Age > 32 {
		ageModifier = 0.7
	}

	return int(float64(baseValue) * ageModifier)
}

func (p *Player) IsDefender() bool {
	return p.Position == PositionCB || p.Position == PositionLB || p.Position == PositionRB
}

func (p *Player) IsMidfielder() bool {
	return p.Position == PositionCDM || p.Position == PositionCM || p.Position == PositionCAM
}

func (p *Player) IsAttacker() bool {
	return p.Position == PositionLW || p.Position == PositionRW || p.Position == PositionST
}

File: flux/apps/games/futboss-ai/packages/shared/src/types/player.ts
// FutBoss AI - Player Types
// Author: Bruno Lucena (bruno@lucena.cloud)

export type Position = 'GK' | 'CB' | 'LB' | 'RB' | 'CDM' | 'CM' | 'CAM' | 'LW' | 'RW' | 'ST';
export type Temperament = 'calm' | 'explosive' | 'calculated';
export type PlayStyle = 'offensive' | 'defensive' | 'balanced';

export interface PlayerAttributes {
  speed: number;
  strength: number;
  stamina: number;
  finishing: number;
  passing: number;
  dribbling: number;
  defense: number;
  intelligence: number;
  aggression: number;
  leadership: number;
  creativity: number;
}

export interface PlayerPersonality {
  temperament: Temperament;
  playStyle: PlayStyle;
}

export interface Player {
  id: string;
  name: string;
  position: Position;
  nationality: string;
  age: number;
  attributes: PlayerAttributes;
  personality: PlayerPersonality;
  teamId?: string;
  price: number;
  isListed: boolean;
  createdAt: Date;
}

export interface PlayerCreate {
  name: string;
  position: Position;
  nationality: string;
  age: number;
  attributes?: Partial<PlayerAttributes>;
  personality?: Partial<PlayerPersonality>;
}

export function calculateOverall(attrs: PlayerAttributes): number {
  const values = Object.values(attrs);
  return Math.round(values.reduce((a, b) => a + b, 0) / values.length);
}

export function calculateMarketValue(player: Player): number {
  const baseValue = calculateOverall(player.attributes) * 10;
  let ageModifier = 1.0;
  
  if (player.age < 23) {
    ageModifier = 1.3;
  } else if (player.age > 32) {
    ageModifier = 0.7;
  }
  
  return Math.round(baseValue * ageModifier);
}

export function isDefender(position: Position): boolean {
  return ['CB', 'LB', 'RB'].includes(position);
}

export function isMidfielder(position: Position): boolean {
  return ['CDM', 'CM', 'CAM'].includes(position);
}

export function isAttacker(position: Position): boolean {
  return ['LW', 'RW', 'ST'].includes(position);
}

